### PR TITLE
Fix UnitTest vm errors

### PR DESCRIPTION
**Pull request by rainbreak**
@MrChico

### CODE REVIEW

File: src/hevm/src/EVM/UnitTest.hs [hunk 1]
```diff
@@ -167,20 +167,23 @@ execTest UnitTestOptions { .. } method args = do
...
-     Stepper.note "Checking whether assertions failed"
-     res <- Stepper.execFullyOrFail >>= Stepper.decode AbiBoolType
-     let AbiBool failed = res
-     -- Return true if the test was successful
-     pure (shouldFail == (bailed || failed))
+  if bailed then
+    pure (False)
+  else do
+    -- Decide whether the test is supposed to fail or succeed
+    let shouldFail = "testFail" `isPrefixOf` method
```
Comment (MrChico) [inline src/hevm/src/EVM/UnitTest.hs:174]:
```suggestion
   -- Decide whether the test is supposed to fail or succeed
  let shouldFail = "testFail" `isPrefixOf` method
  if bailed then
    pure shouldFail
  else do

```

Comment (MrChico) [reply src/hevm/src/EVM/UnitTest.hs:174]: do you agree?

File: src/hevm/src/EVM/UnitTest.hs [hunk 2]
```diff
@@ -548,6 +550,18 @@ failOutput vm dapp UnitTestOptions { .. } testName = mconcat
       _       -> indentLines 2 (showTraceTree dapp vm)
   ]
 
+bailOutput :: Stepper.Failure -> VM -> DappInfo -> UnitTestOptions -> Text -> Text
```
Comment (MrChico) [inline src/hevm/src/EVM/UnitTest.hs:553]: If you let `failOutput` take a `Maybe Stepper.Failure` we could merge `bailOutput` and `failOutput` and get rid of this copypasta

File: src/hevm/src/EVM/UnitTest.hs (hunk 1 above)
Comment (rainbreak) [reply src/hevm/src/EVM/UnitTest.hs:174]: I don't think so, we shouldn't be expecting vm failures with `testFailBlah`. We could perhaps invent a new category of `testBailBlah`.

File: src/hevm/src/EVM/UnitTest.hs (hunk 2 above)
Comment (rainbreak) [reply src/hevm/src/EVM/UnitTest.hs:553]: We could, but I don't think we're done iterating on this area yet so imo it's a bit more flexible to keep separate until we have some bigger vision for rewriting the whole thing.

File: src/hevm/src/EVM/UnitTest.hs (hunk 1 above)
Comment (MrChico) [reply src/hevm/src/EVM/UnitTest.hs:174]: but reverts are vm failures, as are INVALID opcode or anything. The following test should pass:
```
function testFail() {
  require(false);
}
```
This is the current semantics (or, well, before the bug recently introduced).

Comment (rainbreak) [reply src/hevm/src/EVM/UnitTest.hs:174]: Good point. We have

```
ContractNotFound
DecodingError
VMFailed
```

so how about only applying this treatement to `VMFailed` and `DecodingError` and `ContractNotFound` stay as always-fail?

Comment (rainbreak) [reply src/hevm/src/EVM/UnitTest.hs:174]: `ContractNotFound` can be removed actually, it isn't used and we already have `VMError NoSuchContract`. So all we're actually talking about is `DecodingError`.

Comment (rainbreak) [reply src/hevm/src/EVM/UnitTest.hs:174]: Ok I can see that your suggestion will do that now anyway, as `DecodingError` is only ever raised from this particular block of code.

Comment (MrChico) [reply src/hevm/src/EVM/UnitTest.hs]: Yeah. tbh, keeping the decoding inside the stepper script feels a little too much and I could see `Stepper.Failure` being removed completely in favor of just VMError. Then we'd just return false here if the decoding was unsuccessful.

Comment (MrChico) [reply src/hevm/src/EVM/UnitTest.hs]: but that could also be postponed to keep this focused

**Comment by MrChico** [comment]
fixes #420

**Pull request merged**